### PR TITLE
improved JWT claim based routing

### PR DESCRIPTION
**Pull request by wulianglongrd**
issue: [#44228 ](https://github.com/istio/istio/issues/44228#issuecomment-1500775129)

support using `[]` as a separator for nested claim names. This has two benefits:
1. A claim name that includes a domain name can be used to prevent collisions.
2. Same usage experience as AuthorizationPolicy `request.auth.claims` [Condition](https://istio.io/latest/docs/reference/config/security/conditions/).

### CODE REVIEW

**Comment by wulianglongrd** [comment]
@aryan16 Can you please take a look? Not sure if you've noticed this PR since there hasn't been any replies.

**Comment by wulianglongrd** [comment]
closed as no feedback was received.

**Comment by aryan16** [comment]
This would be a breaking change for the existing users as we are moving from `[]` to `.`
Can you update this PR to support both the cases.

**Comment by aryan16** [comment]
Add integ tests as well - https://github.com/istio/istio/blob/master/tests/integration/pilot/common/routing.go#L3283

File: tests/common/jwt/jwt_token.go [hunk 1]
```diff
@@ -94,6 +94,28 @@ const (
...
+	//    }
+	//  },
+	//  "test-issuer-1@istio.io/simple": "valueC"
+	//}
+	// Generated by: python3 security/tools/jwt/samples/gen-jwt.py tests/common/jwt/key.pem -jwks=tests/common/jwt/jwks.json
+	// --expire=3153600000 --iss=test-issuer-1@istio.io --sub=sub-1
+	// --claims=test-issuer-1@istio.io/simple:valueC
+	// --nestedkey=test-issuer-1@istio.io/nested --nestedclaim key1 valueC
+	// nolint: lll
+	TokenIssuer1WithCollisionResistantName = "eyJhbGciOiJSUzI1NiIsImtpZCI6InRUX3c5TFJOclk3d0phbEdzVFlTdDdydXRaaTg2R3Z5YzBFS1I0Q2FRQXciLCJ0eXAiOiJKV1QifQ.eyJleHAiOjQ4MzcyMzE2NTcsImlhdCI6MTY4MzYzMTY1NywiaXNzIjoidGVzdC1pc3N1ZXItMUBpc3Rpby5pbyIsInN1YiI6InN1Yi0xIiwidGVzdC1pc3N1ZXItMUBpc3Rpby5pby9uZXN0ZWQiOnsia2V5MSI6InZhbHVlQyIsIm5lc3RlZC0yIjp7ImtleTEiOiJ2YWx1ZUMifX0sInRlc3QtaXNzdWVyLTFAaXN0aW8uaW8vc2ltcGxlIjoidmFsdWVDIn0.kiXEekepka92nbKNZoPrjx4oeg5OI_ILROfZ9RR02tPsaXmVAEXzfsd5foend_seEru85oLXwiBM8uFJgn8hjE9oD4wXs-xGJHpgTogBeiuzIkrWN0o6iD03tYHFKCTUY8O5B5-9NoyYiEL5pNUqPd16VwfMRg_8qaOZPBVqES-mQED8xVL_UXOlkVQEsb2QUZD_9WfkzO1JRk8m1b6gxxxN2Its1jUB-F667yUQ5tUcYbuaus5yG5sWwpAJTXWzakC2Qq6NfB5hV2fwOsVSLUvH6B7sBq6dz6W03-xFH58j_aAAF0ceJ_LsX63hqZfUOeIcbKCRFoe6Ou8cw5sRtw"
```
Comment (aryan16) [inline tests/common/jwt/jwt_token.go:117]: why do you need to create a new token, we can use the existing tokens? We already have a token with nested claims.

Comment (wulianglongrd) [reply tests/common/jwt/jwt_token.go:117]: > "test-issuer-1@istio.io/simple": "valueC"
> "test-issuer-1@istio.io/nested": {...}

The token's claim name contains special characters `@`, `.` and `/`, which are the scenarios we need to test.

Comment (wulianglongrd) [reply tests/common/jwt/jwt_token.go:117]: We don't have this kind of token now.

**Comment by wulianglongrd** [comment]
> Add integ tests as well - https://github.com/istio/istio/blob/master/tests/integration/pilot/common/routing.go#L3283

Added, thanks.

**Review by aryan16 — APPROVED** [review]

**Review by hanxiaop — APPROVED** [review]

**Review by howardjohn — APPROVED** [review]

**Pull request merged**